Bind 9.17.9 and 9.11.22 Support

julek-wolfssl opened the pull request:
- Add `--enable-bind` configuration option
- New compatibility API:
  - `RSA_get0_crt_params`
  - `RSA_set0_crt_params`
  - `RSA_get0_factors`
  - `RSA_set0_factors`
  - `RSA_test_flags`
  - `HMAC_CTX_get_md`
  - `EVP_MD_block_size`
  - `EC_KEY_check_key`
  - `o2i_ECPublicKey`
  - `DH_get0_key`
  - `DH_set0_key`
- Calling `EVP_MD_CTX_cleanup` on an uninitialized `EVP_MD_CTX` structure is no longer an error
- `DH_generate_parameters` and `DH_generate_parameters_ex` has been implemented

julek-wolfssl committed:
Bind 9.17.9 Support

- Add `--enable-bind` configuration option
- New compatibility API:
  - `RSA_get0_crt_params`
  - `RSA_set0_crt_params`
  - `RSA_get0_factors`
  - `RSA_set0_factors`
  - `RSA_test_flags`
  - `HMAC_CTX_get_md`
  - `EVP_MD_block_size`
  - `EC_KEY_check_key`
  - `o2i_ECPublicKey`
  - `DH_get0_key`
  - `DH_set0_key`
- Calling `EVP_MD_CTX_cleanup` on an uninitialized `EVP_MD_CTX` structure is no longer an error
- `DH_generate_parameters` and `DH_generate_parameters_ex` has been implemented

SparkiDev commented on configure.ac:
```diff
@@ -4272,6 +4284,13 @@ then
+then
+    AM_CFLAGS="$AM_CFLAGS -DWOLFSSL_BIND"
+    AM_CFLAGS="$AM_CFLAGS -DWOLFSSL_AES_DIRECT -DHAVE_AES_ECB -DWOLFSSL_DES_ECB"
+    AM_CFLAGS="$AM_CFLAGS -DWOLFSSL_SHA224 -DWOLFSSL_SHA384 -DWOLFSSL_SHA512"
```
Would be better if enabled_sha224="yes" and so on.
This way any other effects of having it enabled are also happening.

SparkiDev commented on wolfcrypt/src/evp.c:
```diff
@@ -2354,6 +2354,18 @@ static enum wc_HashType wolfSSL_EVP_md2macType(const WOLFSSL_EVP_MD *md)
+{
+    const struct s_ent *ent ;
+
+    for( ent = md_tbl; ent->name != NULL; ent++) {
```
Space between 'for' and '('.
No space between '(' and 'ent'.

SparkiDev replied:
Fix function above too.

julek-wolfssl replied:
Fixed.

SparkiDev commented on wolfcrypt/src/evp.c:
```diff
@@ -2354,6 +2354,18 @@ static enum wc_HashType wolfSSL_EVP_md2macType(const WOLFSSL_EVP_MD *md)
+    const struct s_ent *ent ;
+
+    for( ent = md_tbl; ent->name != NULL; ent++) {
+        if(ent->macType == type) {
```
Space after if.

SparkiDev commented on wolfssl/openssl/crypto.h:
```diff
@@ -26,19 +26,27 @@
+
+typedef struct WOLFSSL_CRYPTO_THREADID {
+    int dummy;
+}WOLFSSL_CRYPTO_THREADID;
```
Space after '}'.

julek-wolfssl replied:
Fixed.

SparkiDev commented on wolfssl/openssl/crypto.h:
```diff
@@ -26,19 +26,27 @@
+typedef struct crypto_threadid_st   CRYPTO_THREADID;
+
+typedef struct CRYPTO_EX_DATA            CRYPTO_EX_DATA;
+typedef void (CRYPTO_free_func)(void*parent, void*ptr, CRYPTO_EX_DATA *ad, int idx,
```
A space after each '*'.

julek-wolfssl replied:
Fixed.

SparkiDev commented on wolfssl/openssl/hmac.h:
```diff
@@ -35,9 +35,17 @@
+    word32  save_ipad[WC_HMAC_BLOCK_SIZE  / sizeof(word32)];  /* same block size all*/
+    word32  save_opad[WC_HMAC_BLOCK_SIZE  / sizeof(word32)];
+} WOLFSSL_HMAC_CTX;
+
```
Don't have includes below structure declaration.

julek-wolfssl replied:
This is necessary because the following includes need to have this struct defined. Moving this after the include results in:
```
In file included from /usr/local/include/wolfssl/openssl/objects.h:30,
                 from /usr/local/include/wolfssl/openssl/evp.h:954,
                 from /usr/local/include/wolfssl/openssl/hmac.h:41,
                 from hmac.c:12:
/usr/local/include/wolfssl/ssl.h:4233:35: error: unknown type name ‘WOLFSSL_HMAC_CTX’; did you mean ‘WOLFSSL_SHA_CTX’?
 4233 |     WOLFSSL_EVP_CIPHER_CTX *ectx, WOLFSSL_HMAC_CTX *hctx, int enc);
      |                                   ^~~~~~~~~~~~~~~~
      |                                   WOLFSSL_SHA_CTX
/usr/local/include/wolfssl/ssl.h:4234:69: error: unknown type name ‘ticketCompatCb’
 4234 | WOLFSSL_API int wolfSSL_CTX_set_tlsext_ticket_key_cb(WOLFSSL_CTX *, ticketCompatCb);
      |                                                                     ^~~~~~~~~~~~~~
```

SparkiDev commented on src/ssl.c:
```diff
@@ -46909,35 +47063,73 @@ void wolfSSL_CRYPTO_cleanup_all_ex_data(void){
+    }
+
+    key = (DhKey*)dh->internal;
+    if (mp_set_int(&key->g, generator) != MP_OKAY) {
```
This doesn't do anything.
mp_init_multi is called with dh->g and this leaks with integer.c.

julek-wolfssl replied:
Right. I tried writing a function to set the generator and check its validity using Appendix A.2.2 of FIPS 186-4 but from testing it looks like only numbers generated from A.2.1 satisfy that condition. I opted to just ignore the input generator parameter and keep the one generated by wolfSSL.

julek-wolfssl committed:
Code review and mp_int memory leak fixes

julek-wolfssl committed:
WIP

julek-wolfssl committed:
Fix race condition with RsaKey

When RsaKey is shared and RsaPublicEncryptEx is called simultaneously by multiple threads, the key->state may be incorrectly set in some threads. This side-steps the state logic when building for bind9.

kaleb-himes commented:
retest this please

julek-wolfssl committed:
dot system test passed

julek-wolfssl commented:
This PR is dependent on https://github.com/wolfSSL/wolfssl/pull/4012 for the OpenSSL compatible defaults.

julek-wolfssl committed:
Bind 9.11.22

julek-wolfssl committed:
Fix DSA signature length

The length of the DSA signature is 40 bytes for N=160 but 64 bytes for N=256. New enum values are added for better clarity.

julek-wolfssl committed:
Rebase fixes

SparkiDev commented on src/ssl.c:
```diff
@@ -54519,12 +55028,48 @@ WOLFSSL_STRING *wolfSSL_TXT_DB_get_by_index(WOLFSSL_TXT_DB *db, int idx,
  * END OF TXT_DB API
  ******************************************************************************/
 
+void wolfSSL_DH_get0_key(const WOLFSSL_DH *dh,
```
Move this to be with other DH APIs.

julek-wolfssl replied:
Moved to other API's.

SparkiDev commented on src/ssl.c:
```diff
@@ -54519,12 +55064,13 @@ WOLFSSL_STRING *wolfSSL_TXT_DB_get_by_index(WOLFSSL_TXT_DB *db, int idx,
  * END OF TXT_DB API
  ******************************************************************************/
 
+#endif
```
Move #endif back.
The #ifdef is inside "START OF"

julek-wolfssl replied:
Moved.

JacobBarthelmeh commented:
reassigning for merge conflicts after recent merged PRs

haydenroche5 commented:
@julek-wolfssl can you resolve these merge conflicts? Looking to get this merged so that my NTP PR can go through, which depends on some of the DSA additions you've made here.

julek-wolfssl commented:
2 days since the last time I fixed conflicts...
![image](https://user-images.githubusercontent.com/49391366/126678062-b6fe858e-96f7-49f7-a676-b038610c97be.png)
I'm on it. Looks like some big conflicts in `evp.c`.

julek-wolfssl committed:
Fix merge conflict resolution in ECC_populate_EVP_PKEY

julek-wolfssl commented:
@dgarske I modified some of the work you recently did with PKCS8 in the EVP layer. Could you take a look?

dgarske commented on src/ssl.c:
```diff
@@ -40246,6 +40707,15 @@ int wolfSSL_EC_KEY_LoadDer_ex(WOLFSSL_EC_KEY* key, const unsigned char* derBuf,
+        idx = 0;
+        ret = ToTraditionalInline(derBuf, &idx, (word32)derSz);
+        if (ret > 0) {
+            ret = wc_EccPrivateKeyDecode(derBuf, &idx,
```
The `wc_EccPrivateKeyDecode` handles PKCS8 format as well now. Do you need the `ToTraditionalInline` check?

julek-wolfssl replied:
Good catch. Removed as its now redundant.

julek-wolfssl committed:
Second `wc_EccPrivateKeyDecode` not needed now that it supports PKCS8

dgarske reviewed (approved):
Sorry @julek-wolfssl , one more conflict due to #4229. Otherwise looks great.

julek-wolfssl committed:
More rebase fixes

julek-wolfssl commented:
@dgarske conflicts fixed. Hopefully that's the last of 'em.
